### ISSUE TITLE
job-summary heuristics use case: Lustre striping

### ISSUE THREAD

**Issue by shanedsnyder** [darshan-hpc/darshan#444]
We have recently discussed adding some heuristics to our job-summary tool to automatically detect certain behaviors in users' Darshan logs and to "grade" users on how they perform on a given heuristic. The idea is to automatically detect good/bad practices in user I/O workloads and to notify them about how good/bad they are doing.

Here, we start to flesh out a heuristic related to improperly set Lustre striping parameters. A Darshan log for this example and a brief README can be found here: https://github.com/darshan-hpc/darshan-logs/tree/main/imbalanced-io

## Lustre striping case

This heuristic would attempt to identify mismatches between Lustre striping parameters and application file access patterns. Specifically, we'd at least like to identify circumstances where users are using a single stripe for large, shared files -- this leads to severely imbalanced workloads for MPI-IO collective workloads, as outlined in #443, where a single aggregator process is selected for doing I/O to a single Lustre stripe server. I found a presentation with some more details on Lustre and MPI-IO interactions if interested (see slide 13 in particular): http://www.archer.ac.uk/training/course-material/2019/03/io-imp/Slides/L02-Cray-IO-Software.pdf

Folks with a better understanding of Lustre striping and MPI-IO behavior might have more opinions on more general Lustre heuristics, but identifying large-scale collective accesses going through 1 (or maybe some other computable threshold, perhaps using file size along with stripe size and width?) Lustre server (OST) does seem like something obvious to catch that shouldn't be too difficult.

As an example, we can see the Lustre striping parameters here for a Lustre file from the above referenced Darshan log:

```
LUSTRE  -1      15708535418621378501    LUSTRE_OSTS     56      /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
LUSTRE  -1      15708535418621378501    LUSTRE_MDTS     4       /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
LUSTRE  -1      15708535418621378501    LUSTRE_STRIPE_OFFSET    0       /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
LUSTRE  -1      15708535418621378501    LUSTRE_STRIPE_SIZE      1048576 /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
LUSTRE  -1      15708535418621378501    LUSTRE_STRIPE_WIDTH     1       /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
LUSTRE  -1      15708535418621378501    LUSTRE_OST_ID_0 29      /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
```

The key counter is the `LUSTRE_STRIPE_WIDTH`, which is set to 1 for this file, indicating it's stored on a single Lustre server (OST). Consider the POSIX counters for this file:

```
POSIX   -1      15708535418621378501    POSIX_BYTES_READ        52939424612     /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
POSIX   -1      15708535418621378501    POSIX_BYTES_WRITTEN     52938395468     /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
POSIX   -1      15708535418621378501    POSIX_MAX_BYTE_READ     52938398983     /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
POSIX   -1      15708535418621378501    POSIX_MAX_BYTE_WRITTEN  52938398983     /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
```

Clearly, a lot of read/write activity is performed to this file (illustrated is total read/write volume, as well as maximum file offset read/wrote, which is a reasonable proxy for file size). Also consider:

```
POSIX   -1      15708535418621378501    POSIX_FASTEST_RANK      465     /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
POSIX   -1      15708535418621378501    POSIX_FASTEST_RANK_BYTES        2072    /lus/theta-fs0/3981085427       /lus/theta-fs0  lust
re
POSIX   -1      15708535418621378501    POSIX_SLOWEST_RANK      0       /lus/theta-fs0/3981085427       /lus/theta-fs0  lustre
POSIX   -1      15708535418621378501    POSIX_SLOWEST_RANK_BYTES        105876790000    /lus/theta-fs0/3981085427       /lus/theta-f
s0  lustre
```

This demonstrates the massive I/O imbalance triggered by the Lustre stripe settings: almost all I/O is going through rank 0.

Metric values:

- This is tricky, because while we could probably say that using a single stripe for large-scale collective I/O is the worst-case scenario (red score), it's not true that larger and larger stripe counts will lead to better and better scores (i.e., not a linear relationship, there's likely a sweet spot, with too many OSTs also being not ideal).

Gotchas:

- Note that darshan-runtime no longer captures `LUSTRE_OSTS`, so if we would rely on that value for this heuristic, we'd need to revisit our runtime code to see how we can capture this value again.

**Comment by carns** [comment]
I would suggest this grading scale as a straw man starting point:

- no lustre data, or none of the heuristics below applicable: gray
- > 64 GiB of data in a globally shared (rank == -1) file with stripe_width == 1: red
- > 64 GiB of data in a globally shared (rank == -1) file with stripe_width/OSTs < 10%: yellow
- > 64 GiB of data in a globally shared (rank == -1) file with stripe_width/OSTs > 10%: green

The color for the job as a whole would be the worst heuristic score out of the files accessed by the application.  The annotation with the score would say something like (in the yellow case, for example) "Detected at least one large globally shared file that used less than 10% of available Lustre storage targets".  Maybe a footnote pointing to a longer form text box that goes into more detail at the very end of the report, so we don't visually clutter the heuristic list?

**Comment by carns** [comment]
Oh, maybe give the count of files that triggered the heuristic (and name one of them as an example?) so that we can tell the difference between a job that did this on one file and a job that did this on 1,000?

**Comment by nawtrey** [comment]
I'm going to start drafting a function for this. 

I did want to double check with @shanedsnyder that this and #443 both use the same log file, correct? At least for the initial testing.

**Comment by nawtrey** [comment]
> Oh, maybe give the count of files that triggered the heuristic (and name one of them as an example?) so that we can tell the difference between a job that did this on one file and a job that did this on 1,000?

For a log with 1000 files that triggered this heuristic, would I see 1000 `POSIX` records with `rank == -1`? So far all I've seen are  logs with a single record with `rank == -1`.

**Comment by shanedsnyder** [comment]
> I'm going to start drafting a function for this.
> 
> I did want to double check with @shanedsnyder that this and #443 both use the same log file, correct? At least for the initial testing.

That's right, same log file demonstrates each issue. Really this issue is a special case of #443, but they probably do make sense as separate heuristics (the other one just giving insight into how I/O is balanced across app. processes, this one for detecting Lustre striping issues specifically).
